File: BluetoothTest/app/src/main/java/com/example/bluetoothtest/activities/MainActivity.java
package com.example.bluetoothtest.activities;

import android.content.Intent;
import android.content.res.Resources;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.PopupWindow;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.example.bluetoothtest.R;
import com.example.bluetoothtest.controllerData.Dpad;

import java.io.File;
import java.io.PrintWriter;
import java.util.Collections;
import java.util.Scanner;
import java.util.Vector;

//TODO allow for no selection available
public class MainActivity extends AppCompatActivity {
    private EditText editText;
    private String oldName;
    private ImageButton rename_btn;

    private RelativeLayout relativeLayout;

    private Vector<String> layouts;

    private View decorView;

    private PopupWindow popupWindow = new PopupWindow(); //so we can access it easily

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        editText = findViewById(R.id.editText);
        rename_btn = findViewById(R.id.rename_btn);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        getSupportActionBar().setDisplayHomeAsUpEnabled(false);
        getSupportActionBar().setTitle("");

        File path;
        path = new File(getFilesDir() + "/macros/");

        if (!path.exists()) {
            path.mkdir();
        }

        if (path.listFiles().length==0) createMacros();

        path = new File(getFilesDir() + "/layouts/");
        layouts = new Vector<>();
        //Log.d("ZZZ",path.toString());

        if (!path.exists()) {
            path.mkdir();
        }

        if (path.listFiles().length==0) createLayouts();

        for (File i : path.listFiles()) {
            //Log.d("ZZZ","File found: "+i.getName());
            layouts.add(i.getName());
        }

        Collections.sort(layouts);

        //add them to edittext
        editText.setText(layouts.get(0));

        editText.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                return false;
            }
        });

        relativeLayout = findViewById(R.id.relative_layout);
        LinearLayout.LayoutParams layoutParams = (LinearLayout.LayoutParams) relativeLayout.getLayoutParams();
        layoutParams.height = Resources.getSystem().getDisplayMetrics().heightPixels / 2 + 40;
        layoutParams.width = Resources.getSystem().getDisplayMetrics().widthPixels / 2 + 40;
        relativeLayout.setLayoutParams(layoutParams);

        //code to make app bigger
        decorView = getWindow().getDecorView();
        decorView.setOnSystemUiVisibilityChangeListener(new View.OnSystemUiVisibilityChangeListener() {
            @Override
            public void onSystemUiVisibilityChange(int visibility) {
                if (visibility == 0) {
                    setHighVisibility();
                }
            }
        });
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        if (hasFocus) {
            setHighVisibility();
        }
    }

    private void setHighVisibility() { //hide nav bar and make app fullscreen
        decorView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE |
                View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY |
                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN |
                View.SYSTEM_UI_FLAG_FULLSCREEN |
                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN |
                View.SYSTEM_UI_FLAG_HIDE_NAVIGATION |
                View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
        );
    }

    @Override
    public void onResume() {
        super.onResume();
        fillPreview();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.top_nav_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.settingsBtn:
                startActivity(new Intent(this.getApplicationContext(), SettingsActivity.class));
                break;
        }


        return super.onOptionsItemSelected(item);
    }

    public void start(View view) {
        if (editText.isFocusable()) {
            Toast.makeText(this, "Finish editing first!", Toast.LENGTH_SHORT).show();
            return;
        }

        if (editText.isFocusable()) {
            Toast.makeText(this, "Finish editing first!", Toast.LENGTH_SHORT).show();
            return;
        }

        Intent intent = new Intent(this, ControllerActivity.class);
        intent.putExtra("preset", editText.getText().toString());
        startActivity(intent);
    }

    public void edit(View view) {
        if (editText.isFocusable()) {
            Toast.makeText(this, "Finish editing first!", Toast.LENGTH_SHORT).show();
            return;
        }

        Intent intent = new Intent(this, EditActivity.class);
        intent.putExtra("preset", editText.getText().toString());
        startActivity(intent);
    }

    public void macro(View view) {
        if (editText.isFocusable()) {
            Toast.makeText(this, "Finish editing first!", Toast.LENGTH_SHORT).show();
            return;
        }

        Intent intent = new Intent(this, MacroActivity.class);
        startActivity(intent);
    }

    public void inflate(View view) {
        if (editText.isFocusable()) return; //dont open if user is trying to rename a file

        if (popupWindow.isShowing()) {
            LinearLayout linearLayout = popupWindow.getContentView().findViewById(R.id.linear_layout);
            linearLayout.removeAllViews();
            for (int i = 0; i < layouts.size(); i++) {
                TextView textView = new TextView(this);
                textView.setMinHeight(0);
                textView.setMinWidth(0);
                textView.setText(layouts.get(i));
                textView.setTextSize(20);

                linearLayout.addView(textView, 0);
            }
        } else {
            LayoutInflater layoutInflater
                    = (LayoutInflater) getBaseContext()
                    .getSystemService(LAYOUT_INFLATER_SERVICE);
            View popupView = layoutInflater.inflate(R.layout.popup_choose, null);


            LinearLayout linearLayout = popupView.findViewById(R.id.linear_layout);
            linearLayout.setMinimumWidth(editText.getWidth());
            for (int i = 0; i < layouts.size(); i++) {
                TextView textView = new TextView(this);
                textView.setMinHeight(0);
                textView.setMinWidth(0);
                textView.setText(layouts.get(i));
                textView.setTextSize(20);

                textView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        editText.setText(((TextView) v).getText().toString());
                        fillPreview();
                        popupWindow.dismiss();
                    }
                });

                linearLayout.addView(textView, i);
            }

            popupWindow = new PopupWindow(
                    popupView,
                    editText.getWidth(),
                    ViewGroup.LayoutParams.WRAP_CONTENT);

            popupWindow.setFocusable(true);
            popupWindow.update();
            popupWindow.showAsDropDown(view);
        }
    }

    public void add(View view) {
        if (editText.isFocusable()) {
            Toast.makeText(this, "Finish editing first!", Toast.LENGTH_SHORT).show();
            return;
        }

        File file;
        int index = 1;

        while (true) {
            file = new File(getFilesDir() + "/layouts/Layout " + index);
            if (!file.exists()) break;
            index++;
        }

        try {
            file.createNewFile();
        } catch (Exception e) {
            Toast.makeText(this, "Something went wrong", Toast.LENGTH_SHORT).show();
            return;
        }

        layouts.add(file.getName());
        Collections.sort(layouts);

        editText.setText(file.getName());
        fillPreview();
    }

    public void del(View view) {
        if (editText.isFocusable()) {
            Toast.makeText(this, "Finish editing first!", Toast.LENGTH_SHORT).show();
            return;
        }

        if (layouts.size() == 1) {
            Toast.makeText(this, "There must always be at least 1 layout!", Toast.LENGTH_SHORT).show();
            return;
        }

        File file = new File(getFilesDir() + "/layouts/" + editText.getText());

        try {
            file.delete();
        } catch (Exception e) {
            Toast.makeText(this, "Something went wrong", Toast.LENGTH_SHORT).show();
            return;
        }

        for (int i = 0; i < layouts.size(); i++) {
            if (layouts.get(i).equals(file.getName())) {
                layouts.remove(i);
                editText.setText(layouts.get(Math.min(i, layouts.size() - 1)));
                break;
            }
        }
        fillPreview();
    }

    public void rename(View view) {
        if (editText.isFocusable()) { //check if renaming is valid
            String newName = editText.getText().toString();

            if (!newName.equals(oldName)) {
                File file = new File(getFilesDir() + "/layouts/" + newName);
                if (file.exists()) {
                    Toast.makeText(this, "Invalid file name", Toast.LENGTH_SHORT).show();
                    return;
                }

                for (int i = 0; i < layouts.size(); i++) {
                    if (layouts.get(i).equals(oldName)) layouts.set(i, newName);
                }

                File file2 = new File(getFilesDir() + "/layouts/" + oldName);
                file2.renameTo(file);

                Collections.sort(layouts);
            }

            editText.setFocusable(false);
            editText.setFocusableInTouchMode(false);
            rename_btn.setImageResource(R.drawable.ic_rename);
        } else { //allow user to edit
            oldName = editText.getText().toString();

            editText.setFocusable(true);
            editText.setFocusableInTouchMode(true);

            rename_btn.setImageResource(R.drawable.ic_tick);
        }
    }

    //wow we are gonna show a preview bcas thats cool
    private void fillPreview() {
        relativeLayout.removeAllViews();
        try {
            File file = new File(getFilesDir() + "/layouts/" + editText.getText().toString());
            Scanner scanner = new Scanner(file);

            while (scanner.hasNextLine()) {
                String[] args = scanner.nextLine().split("\0");

                switch (args[0]) {
                    case "Btn":
                        Btn(Integer.parseInt(args[3]), Integer.parseInt(args[4]), Integer.parseInt(args[5]), Integer.parseInt(args[6]));
                        break;

                    case "Dpad":
                        Dpad(Integer.parseInt(args[5]), Integer.parseInt(args[6]), Integer.parseInt(args[7]));
                        break;

                    case "Macro":
                        Macro(Integer.parseInt(args[4]), Integer.parseInt(args[5]), Integer.parseInt(args[6]), Integer.parseInt(args[7]));
                        break;
                }
            }

            scanner.close();
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(this, "File corrupted >.<", Toast.LENGTH_SHORT).show();
            finish();
        }
    }

    private void Btn(int height, int width, int marginTop, int marginLeft) {
        height /= 2;
        width /= 2;
        marginTop /= 2;
        marginLeft /= 2;

        Button btn = new Button(this);

        btn.setHeight(height);
        btn.setWidth(width);
        btn.setMinimumHeight(0);
        btn.setMinimumWidth(0);

        btn.setBackgroundResource(R.drawable.button_up);

        relativeLayout.addView(btn);

        RelativeLayout.LayoutParams layoutParams = (RelativeLayout.LayoutParams) btn.getLayoutParams();
        layoutParams.leftMargin = marginLeft;
        layoutParams.topMargin = marginTop;
        btn.setLayoutParams(layoutParams);
    }

    private void Dpad(int diameter, int marginTop, int marginLeft) {
        diameter /= 2;
        marginTop /= 2;
        marginLeft /= 2;

        Button btn = new Button(this);

        btn.setHeight(diameter);
        btn.setWidth(diameter);
        btn.setMinimumHeight(0);
        btn.setMinimumWidth(0);

        btn.setBackgroundResource(R.drawable.dpad);

        relativeLayout.addView(btn);

        RelativeLayout.LayoutParams layoutParams = (RelativeLayout.LayoutParams) btn.getLayoutParams();
        layoutParams.leftMargin = marginLeft;
        layoutParams.topMargin = marginTop;
        btn.setLayoutParams(layoutParams);
    }

    private void Macro(int height, int width, int marginTop, int marginLeft) {
        height /= 2;
        width /= 2;
        marginTop /= 2;
        marginLeft /= 2;

        ImageView btn = new ImageView(this);

        btn.setMaxHeight(height);
        btn.setMaxWidth(width);
        btn.setMinimumHeight(height);
        btn.setMinimumWidth(width);

        btn.setBackgroundResource(R.drawable.button_macro);

        btn.setImageResource(R.drawable.ic_m);
        btn.setScaleType(ImageView.ScaleType.FIT_CENTER);
        btn.setAdjustViewBounds(false);

        relativeLayout.addView(btn);

        RelativeLayout.LayoutParams layoutParams = (RelativeLayout.LayoutParams) btn.getLayoutParams();
        layoutParams.leftMargin = marginLeft;
        layoutParams.topMargin = marginTop;
        btn.setLayoutParams(layoutParams);
    }

    //someone suggested that i create presets for users
    private void createLayouts(){
        try {
            File file = new File(getFilesDir() + "/layouts/" + "Layout");
            PrintWriter pw=new PrintWriter(file);

            pw.println("Dpad"+"\0"+"w"+"\0"+"s"+"\0"+"a"+"\0"+"d"+"\0"+getResources().getDisplayMetrics().heightPixels+"\0"+"0"+"\0"+"0");
            pw.println("Btn"+"\0"+"Button"+"\0"+"space"+"\0"+getResources().getDisplayMetrics().heightPixels/2+"\0"+getResources().getDisplayMetrics().heightPixels/2+"\0"+getResources().getDisplayMetrics().heightPixels/2+"\0"+getResources().getDisplayMetrics().heightPixels);
            pw.println("Macro"+"\0"+"Macro"+"\0"+getFilesDir()+"/macros/"+"25268206-7a75-11ea-bc55-0242ac130003"+"\0"+false+"\0"+getResources().getDisplayMetrics().heightPixels/2+"\0"+getResources().getDisplayMetrics().heightPixels/2+"\0"+"0"+"\0"+getResources().getDisplayMetrics().heightPixels);

            pw.flush();
            pw.close();

        }
        catch (Exception e){
            e.printStackTrace();
        }
    }

    private void createMacros(){
        try {
            File file = new File(getFilesDir() + "/macros/" + "25268206-7a75-11ea-bc55-0242ac130003");
            PrintWriter pw=new PrintWriter(file);

            pw.println("Jitter click");
            pw.println("KeyStroke"+"\0"+"m1"+"\0"+true+"\0"+false+"\0"+"367ccb0c-d582-42e0-b004-da3d9d2159e1");
            pw.println("KeyStroke"+"\0"+"m1"+"\0"+false+"\0"+true+"\0"+"367ccb0c-d582-42e0-b004-da3d9d2159e1");
            pw.println("Delay"+"\0"+"200"+"\0"+false+"\0"+false+"\0"+"71089e51-d764-4e28-b25c-2f2d8a8f4800");

            pw.flush();
            pw.close();

        }
        catch (Exception e){
            e.printStackTrace();
        }
    }
}

File: BluetoothTest/app/src/main/java/com/why/controller/activities/MainActivity.kt
package com.why.controller.activities

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.res.Resources
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.view.*
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.why.bluetoothtouchpad2.bluetooth.BluetoothController
import com.why.bluetoothtouchpad2.bluetooth.MouseSender
import com.why.bluetoothtouchpad2.bluetooth.Sender
import com.why.controller.R
import com.why.controller.bluetooth.KeyboardSender
import com.why.controller.bluetooth.Main
import com.why.controller.bluetooth.Main.keyboard
import com.why.controller.bluetooth.Main.mouse
import java.io.File
import java.io.PrintWriter
import java.util.*

//TODO allow for no selection available

//TODO remove setttings
class MainActivity : AppCompatActivity() {
    private var editText: EditText? = null
    private var oldName: String? = null
    private var rename_btn: ImageButton? = null
    private var relativeLayout: RelativeLayout? = null
    private var layouts: Vector<String>? = null
    private var decorView: View? = null
    private var popupWindow = PopupWindow() //so we can access it easily
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        editText = findViewById(R.id.editText)
        rename_btn = findViewById(R.id.rename_btn)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(false)
        supportActionBar!!.title = ""
        var path = File("$filesDir/macros/")
        if (!path.exists()) {
            path.mkdir()
        }
        if (path.listFiles().isNullOrEmpty()) {
            createMacros()
        }
        path = File("$filesDir/layouts/")
        layouts = Vector()
        //Log.d("ZZZ",path.toString());
        if (!path.exists()) {
            path.mkdir()
        }
        if (path.listFiles().isNullOrEmpty()) {
            createLayouts()
        }
        for (i in path.listFiles()!!) {
            //Log.d("ZZZ","File found: "+i.getName());
            layouts!!.add(i.name)
        }
        layouts?.sort()

        //add them to edittext
        editText?.setText(layouts!![0])
        editText?.setOnEditorActionListener { v, actionId, event -> false }
        relativeLayout = findViewById(R.id.relative_layout)
        val layoutParams = relativeLayout?.layoutParams as LinearLayout.LayoutParams
        layoutParams.height = Resources.getSystem().displayMetrics.heightPixels / 2 + 40
        layoutParams.width = Resources.getSystem().displayMetrics.widthPixels / 2 + 40
        relativeLayout?.layoutParams = layoutParams

        //code to make app bigger
        decorView = window.decorView
        decorView!!.setOnSystemUiVisibilityChangeListener { visibility ->
            if (visibility == 0) {
                setHighVisibility()
            }
        }
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            setHighVisibility()
        }
    }

    private fun setHighVisibility() { //hide nav bar and make app fullscreen
        decorView!!.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY or
                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or
                View.SYSTEM_UI_FLAG_FULLSCREEN or
                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or
                View.SYSTEM_UI_FLAG_HIDE_NAVIGATION or
                View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
    }

    public override fun onResume() {
        super.onResume()
        fillPreview()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.top_nav_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.settingsBtn -> startActivity(Intent(this.applicationContext, SettingsActivity::class.java))
        }
        return super.onOptionsItemSelected(item)
    }

    fun start(view: View?) {
        if (editText!!.isFocusable) {
            Toast.makeText(this, "Finish editing first!", Toast.LENGTH_SHORT).show()
            return
        }
        if (editText!!.isFocusable) {
            Toast.makeText(this, "Finish editing first!", Toast.LENGTH_SHORT).show()
            return
        }
        val intent = Intent(this, ControllerActivity::class.java)
        intent.putExtra("preset", editText!!.text.toString())
        startActivity(intent)
    }

    fun edit(view: View?) {
        if (editText!!.isFocusable) {
            Toast.makeText(this, "Finish editing first!", Toast.LENGTH_SHORT).show()
            return
        }
        val intent = Intent(this, EditActivity::class.java)
        intent.putExtra("preset", editText!!.text.toString())
        startActivity(intent)
    }

    fun macro(view: View?) {
        if (editText!!.isFocusable) {
            Toast.makeText(this, "Finish editing first!", Toast.LENGTH_SHORT).show()
            return
        }
        val intent = Intent(this, MacroActivity::class.java)
        startActivity(intent)
    }

    fun inflate(view: View?) {
        if (editText!!.isFocusable) return  //dont open if user is trying to rename a file
        if (popupWindow.isShowing) {
            val linearLayout = popupWindow.contentView.findViewById<LinearLayout>(R.id.linear_layout)
            linearLayout.removeAllViews()
            for (i in layouts!!.indices) {
                val textView = TextView(this)
                textView.minHeight = 0
                textView.minWidth = 0
                textView.text = layouts!![i]
                textView.textSize = 20f
                linearLayout.addView(textView, 0)
            }
        } else {
            val layoutInflater = baseContext
                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            val popupView = layoutInflater.inflate(R.layout.popup_choose, null)
            val linearLayout = popupView.findViewById<LinearLayout>(R.id.linear_layout)
            linearLayout.minimumWidth = editText!!.width
            for (i in layouts!!.indices) {
                val textView = TextView(this)
                textView.minHeight = 0
                textView.minWidth = 0
                textView.text = layouts!![i]
                textView.textSize = 20f
                textView.setOnClickListener { v ->
                    editText!!.setText((v as TextView).text.toString())
                    fillPreview()
                    popupWindow.dismiss()
                }
                linearLayout.addView(textView, i)
            }
            popupWindow = PopupWindow(
                    popupView,
                    editText!!.width,
                    ViewGroup.LayoutParams.WRAP_CONTENT)
            popupWindow.isFocusable = true
            popupWindow.update()
            popupWindow.showAsDropDown(view)
        }
    }

    fun add(view: View?) {
        if (editText!!.isFocusable) {
            Toast.makeText(this, "Finish editing first!", Toast.LENGTH_SHORT).show()
            return
        }
        var file: File
        var index = 1
        while (true) {
            file = File("$filesDir/layouts/Layout $index")
            if (!file.exists()) break
            index++
        }
        try {
            file.createNewFile()
        } catch (e: Exception) {
            Toast.makeText(this, "Something went wrong", Toast.LENGTH_SHORT).show()
            return
        }
        layouts!!.add(file.name)
        layouts!!.sort()
        editText!!.setText(file.name)
        fillPreview()
    }

    fun del(view: View?) {
        if (editText!!.isFocusable) {
            Toast.makeText(this, "Finish editing first!", Toast.LENGTH_SHORT).show()
            return
        }
        if (layouts!!.size == 1) {
            Toast.makeText(this, "There must always be at least 1 layout!", Toast.LENGTH_SHORT).show()
            return
        }
        val file = File(filesDir.toString() + "/layouts/" + editText!!.text)
        try {
            file.delete()
        } catch (e: Exception) {
            Toast.makeText(this, "Something went wrong", Toast.LENGTH_SHORT).show()
            return
        }
        for (i in layouts!!.indices) {
            if (layouts!![i] == file.name) {
                layouts!!.removeAt(i)
                editText!!.setText(layouts!![Math.min(i, layouts!!.size - 1)])
                break
            }
        }
        fillPreview()
    }

    fun rename(view: View?) {
        if (editText!!.isFocusable) { //check if renaming is valid
            val newName = editText!!.text.toString()
            if (newName != oldName) {
                val file = File("$filesDir/layouts/$newName")
                if (file.exists()) {
                    Toast.makeText(this, "Invalid file name", Toast.LENGTH_SHORT).show()
                    return
                }
                for (i in layouts!!.indices) {
                    if (layouts!![i] == oldName) layouts!![i] = newName
                }
                val file2 = File("$filesDir/layouts/$oldName")
                file2.renameTo(file)
                Collections.sort(layouts)
            }
            editText!!.isFocusable = false
            editText!!.isFocusableInTouchMode = false
            rename_btn!!.setImageResource(R.drawable.ic_rename)
        } else { //allow user to edit
            oldName = editText!!.text.toString()
            editText!!.isFocusable = true
            editText!!.isFocusableInTouchMode = true
            rename_btn!!.setImageResource(R.drawable.ic_tick)
        }
    }

    //wow we are gonna show a preview bcas thats cool
    private fun fillPreview() {
        relativeLayout!!.removeAllViews()
        try {
            val file = File(filesDir.toString() + "/layouts/" + editText!!.text.toString())
            val scanner = Scanner(file)
            while (scanner.hasNextLine()) {
                val args = scanner.nextLine().split("\u0000".toRegex()).toTypedArray()
                when (args[0]) {
                    "Btn" -> Btn(args[3].toInt(), args[4].toInt(), args[5].toInt(), args[6].toInt())
                    "Dpad" -> Dpad(args[5].toInt(), args[6].toInt(), args[7].toInt())
                    "Macro" -> Macro(args[4].toInt(), args[5].toInt(), args[6].toInt(), args[7].toInt())
                    "JoyStick" -> JoyStick(args[2].toInt(), args[3].toInt(), args[4].toInt())
                }
            }
            scanner.close()
        } catch (e: Exception) {
            e.printStackTrace()
            Toast.makeText(this, "File corrupted >.<", Toast.LENGTH_SHORT).show()
            finish()
        }
    }

    private fun Btn(height: Int, width: Int, marginTop: Int, marginLeft: Int) {
        var height = height
        var width = width
        var marginTop = marginTop
        var marginLeft = marginLeft
        height /= 2
        width /= 2
        marginTop /= 2
        marginLeft /= 2
        val btn = Button(this)
        btn.height = height
        btn.width = width
        btn.minimumHeight = 0
        btn.minimumWidth = 0
        btn.setBackgroundResource(R.drawable.button_up)
        relativeLayout!!.addView(btn)
        val layoutParams = btn.layoutParams as RelativeLayout.LayoutParams
        layoutParams.leftMargin = marginLeft
        layoutParams.topMargin = marginTop
        btn.layoutParams = layoutParams
    }

    private fun Dpad(diameter: Int, marginTop: Int, marginLeft: Int) {
        var diameter = diameter
        var marginTop = marginTop
        var marginLeft = marginLeft
        diameter /= 2
        marginTop /= 2
        marginLeft /= 2
        val btn = Button(this)
        btn.height = diameter
        btn.width = diameter
        btn.minimumHeight = 0
        btn.minimumWidth = 0
        btn.setBackgroundResource(R.drawable.dpad)
        relativeLayout!!.addView(btn)
        val layoutParams = btn.layoutParams as RelativeLayout.LayoutParams
        layoutParams.leftMargin = marginLeft
        layoutParams.topMargin = marginTop
        btn.layoutParams = layoutParams
    }

    private fun Macro(height: Int, width: Int, marginTop: Int, marginLeft: Int) {
        var height = height
        var width = width
        var marginTop = marginTop
        var marginLeft = marginLeft
        height /= 2
        width /= 2
        marginTop /= 2
        marginLeft /= 2
        val btn = ImageView(this)
        btn.maxHeight = height
        btn.maxWidth = width
        btn.minimumHeight = height
        btn.minimumWidth = width
        btn.setBackgroundResource(R.drawable.button_macro)
        btn.setImageResource(R.drawable.ic_m)
        btn.scaleType = ImageView.ScaleType.FIT_CENTER
        btn.adjustViewBounds = false
        relativeLayout!!.addView(btn)
        val layoutParams = btn.layoutParams as RelativeLayout.LayoutParams
        layoutParams.leftMargin = marginLeft
        layoutParams.topMargin = marginTop
        btn.layoutParams = layoutParams
    }

    private fun JoyStick(diameter: Int, marginTop: Int, marginLeft: Int) {
        var diameter = diameter
        var marginTop = marginTop
        var marginLeft = marginLeft
        diameter /= 2
        marginTop /= 2
        marginLeft /= 2
        val btn = Button(this)
        btn.height = diameter
        btn.width = diameter
        btn.minimumHeight = 0
        btn.minimumWidth = 0
        btn.setBackgroundResource(R.drawable.dpad)
        relativeLayout!!.addView(btn)
        val layoutParams = btn.layoutParams as RelativeLayout.LayoutParams
        layoutParams.leftMargin = marginLeft
        layoutParams.topMargin = marginTop
        btn.layoutParams = layoutParams
    }

    //someone suggested that i create presets for users
    private fun createLayouts() {
        try {
            val file = File("$filesDir/layouts/Layout")
            val pw = PrintWriter(file)
            pw.println("Dpad" + "\u0000" + "w" + "\u0000" + "s" + "\u0000" + "a" + "\u0000" + "d" + "\u0000" + resources.displayMetrics.heightPixels + "\u0000" + "0" + "\u0000" + "0")
            pw.println("Btn" + "\u0000" + "Button" + "\u0000" + "space" + "\u0000" + resources.displayMetrics.heightPixels / 2 + "\u0000" + resources.displayMetrics.heightPixels / 2 + "\u0000" + resources.displayMetrics.heightPixels / 2 + "\u0000" + resources.displayMetrics.heightPixels)
            pw.println("Macro" + "\u0000" + "Macro" + "\u0000" + filesDir + "/macros/" + "25268206-7a75-11ea-bc55-0242ac130003" + "\u0000" + false + "\u0000" + resources.displayMetrics.heightPixels / 2 + "\u0000" + resources.displayMetrics.heightPixels / 2 + "\u0000" + "0" + "\u0000" + resources.displayMetrics.heightPixels)
            pw.flush()
            pw.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun createMacros() {
        try {
            val file = File("$filesDir/macros/25268206-7a75-11ea-bc55-0242ac130003")
            val pw = PrintWriter(file)
            pw.println("Jitter click")
            pw.println("KeyStroke" + "\u0000" + "m1" + "\u0000" + true + "\u0000" + false + "\u0000" + "367ccb0c-d582-42e0-b004-da3d9d2159e1")
            pw.println("KeyStroke" + "\u0000" + "m1" + "\u0000" + false + "\u0000" + true + "\u0000" + "367ccb0c-d582-42e0-b004-da3d9d2159e1")
            pw.println("Delay" + "\u0000" + "200" + "\u0000" + false + "\u0000" + false + "\u0000" + "71089e51-d764-4e28-b25c-2f2d8a8f4800")
            pw.flush()
            pw.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    var m: Sender? = null


    @ExperimentalUnsignedTypes
    @SuppressLint("ClickableViewAccessibility")
    override fun onStart() {
        super.onStart()
        Main.init(this.applicationContext)
        Log.d("ZZZ", "eoiugneoigrng")
        BluetoothController.getSender { hid, device ->
            Log.d("ZZZ", "BL very good")
            Main.connected = true

            val mainHandler = Handler(this.mainLooper)
            mainHandler.post {
                mouse = MouseSender(hid, device)
                keyboard = KeyboardSender(hid, device)
                m = Sender(mouse!!, hid, device)
                Toast.makeText(applicationContext, "Bluetooth is connected!", Toast.LENGTH_LONG).show()

            }

        }

        BluetoothController.getDisconnector {
            val mainHandler = Handler(this.mainLooper)
            mainHandler.post {
                println("byebyebye")
                Main.connected = false
            }
        }


    }
}